swupd/delta.go

```diff
@@ -0,0 +1,160 @@
+	var files []*File
+
+	if manifest == nil {
+		return nil, nil
+	}
+
+	for _, file := range manifest.Files {
+		if file.Version <= manifest.Header.Version ||
+			file.DeltaPeer == nil ||
+			(file.Type != typeFile && file.Type != typeDirectory && file.Type != typeLink) {
+			continue
+		}
+
+		deltaFile := fmt.Sprintf("%s/%d/delta/%d-%d-%s-%s",
+			c.outputDir, file.Version, file.DeltaPeer.Version,
+			file.Version, file.DeltaPeer.Hash, file.Hash)
+
+		// Only add to list if the delta file does not already exist on the
+		// system or in the list
+		if _, err := os.Stat(deltaFile); os.IsNotExist(err) {
+			if found := file.findFileNameInSlice(files); found == nil {
+				// Prepend to list...
+				files = append([]*File{file}, files...)
+			}
+		}
+
+	}
+
+	return files, nil
+}
+
+func createDelta(file *File, statedir string, fromVersion uint32, fromHash Hashval) error {
+	var c config
+	c, err := getConfig(statedir)
+	if err != nil {
+		return err
+	}
+
+	// File without peers cannot have deltas
+	if file.DeltaPeer == nil {
+		return nil
+	}
+
+	// We only support deltas between two regular files for now
+	if file.Type != typeFile || file.DeltaPeer.Type != typeFile {
+		return nil
+	}
+
+	// By definition we cannot make deltas for files that are deleted
+	if file.Status == statusDeleted {
+		return nil
+	}
+
+	// TODO: Feature to support in the future, this is a complex case
+	if file.Type == typeDirectory {
+		return nil
+	}
+
+	newfile := fmt.Sprintf("%s/%d/full/%s", c.imageBase, file.Version, file.Name)
+	original := fmt.Sprintf("%s/%d/full/%s", c.imageBase, fromVersion, file.Name)
+
+	deltafile := fmt.Sprintf("%s/%d/delta/%d-%d-%s-%s", c.outputDir, file.Version, fromVersion, file.Version, fromHash, file.Hash)
+	dotfile := fmt.Sprintf("%s/%d/delta/.%d-%d-%s-%s", c.outputDir, file.Version, fromVersion, file.Version, fromHash, file.Hash)
+	testnewfile := fmt.Sprintf("%s/%d/delta/.%d-%d-%s-%s.testnewfile", c.outputDir, file.Version, fromVersion, file.Version, fromHash, file.Hash)
+
+	if sameFile(file, file.DeltaPeer) {
+		return nil
+	}
+
+	// Shell out to bsdiff for now...false = don't print to screen
+	if err := helpers.RunCommand(false, "bsdiff", original, newfile, dotfile); err != nil {
+		_ = os.Remove(dotfile) // Might have retuned FULLDL
+		return errors.Wrapf(err, "Failed to create delta for %s -> %s", original, newfile)
+	}
+
+	// Check that the delta actually applies correctly
+	if err := helpers.RunCommand(false, "bspatch", original, testnewfile, dotfile); err != nil {
+		return errors.Wrapf(err, "Failed to apply delta %s", deltafile)
+	}
+
+	deltahash, err := Hashcalc(testnewfile)
+	if err != nil {
+		return errors.Wrap(err, "Failed to calculate hash for new file")
+	}
+
+	if !HashEquals(deltahash, file.Hash) {
+		return errors.Wrapf(err, "Delta mismatch: %s -> %s via delta: %s", original, newfile, deltafile)
+	}
+
+	_ = os.Remove(testnewfile)
+
+	// add rename code here
```

Comment:
This may not be necessary depending on rename implementation. Rename code should be setting `file.DeltaPeer` correctly... @icarus-sparry?